docs/apache-airflow/production-deployment.rst

```diff
@@ -222,6 +222,29 @@ you can exchange the Google Cloud Platform identity to the Amazon Web Service id
 which effectively means access to Amazon Web Service platform.
 For more information, see: :ref:`howto/connection:aws:gcp-federation`
 
+
+Applying patches from ``main``
+==============================
+
+On occasion, a user may want to apply a patch from ``main`` which has not yet made it into a release.  Assuming you want to apply a specific PR to one of the official airflow images, you can apply roughly as follows.  It's also possible to apply a specific commit.
+
+.. code-block:: docker
```

Comment:
WDYT about showing both examples explicitly (i.e. patching from PR and from a commit)? I believe the syntax is slightly different but important. 

Have "It's also possible to apply a specific commit." on a new line and show an example of patching from a commit. Do you think seeing both examples would be helpful for beginners (albeit with some healthy of overlap)?

Reply:
right yeah it's a hard one... i thought about that ... i was trying to just be like... here's an example... there are other things you can do too so go google :)  

because.... we already have a _lot_ of docs... and not trying to be an exhaustive 

could patch single commit, could patch diff between two commits... or two tags... or what have you 

e.g. here's the patch between 2.4 and some commit https://github.com/apache/airflow/compare/2.4.0..21c40cd4cc7830bc17a4180d8dd0fa2cc4ed9cea.patch

what do you think's the best balance to strike?

Reply:
Right, I don't feel too strongly about this. I only brought it up because it was specifically mentioned and I kind of wanted it to be there while reading the section. Maybe remove that sentence? I would imagine the top use case would be a PR patch.

If users find this doc and want to add more examples that they found useful, they are welcome to open a PR for those.

Reply:
Oh. Wow. I did not even know we have such instructions. I think we should really target that for "power" users. Any attempt to explain what they should do more precisely and with more examples will inevitably lead to users trying to do more things (resolving conflicts for example or applying two patches at once)  and complaining there are no instructions.

I think more detailed instructions are useful only if we can can maintain them, test and actually make fully "copy&pasteable". One thing that I'd add here is to stress this is just example and add a warning "don't try it at home" if you do not know what you're doing.

It's a bit dangerous to be too specific here.

Reply:
Just as an example - I have a number of examples for Docker "Customising" in https://airflow.apache.org/docs/docker-stack/build.html#examples-of-image-customizing. -but they are all taken from actual scripts that I run from time to time to test if they still work (they take long time to run in CI) and this whole section is for "power users".

Reply:
> Oh. Wow. I did not even know we have such instructions

well i guess we don't yet 😛

Are you saying we shouldn't put it here?  Do you think it belongs in customizing image? I thought it did not because it's not what that page is about.  If there's a better place lemme know.  

At our company, it's not terribly uncommon to help users out by letting them apply a fix quickly... Because sometimes you just need to patch something and release process will always take _some_ time.

Reply:
It's OK to have it. but IMHO 

a) it should be ideally automatically tested that it works (see https://github.com/apache/airflow/blob/main/.github/workflows/ci.yml#L581  where we test image building examples)
b) there should be a note saying "this is just example and you need to know what you are doing" of sorts.

Otherwise it's a question of time to see an issue "I did it this way as described and it did not work, what am I doing wrong?".

Reply:
I think b) is already pretty much there.

Reply:
I'm not sure we really need to "test" this, as the boilderplate itself is pretty darn stable.

I definitely think we should add a warning that patches don't always apply cleanly, and you are on your own in that case.

I almost wonder if we shouldn't even have a real PR number in the example?

Reply:
Seems kindof pointless to add if not in the official documentation.  The intended audience is users.  I'm not too personally invested in this, just, from a pragmatic perspective, this is something that is useful for users of airflow, and it seemed like a useful thing to be able to refer to.  If you really think it shouldn't be in the docs, I'm ok just closing the PR.

Reply:
The more I think about it the more i think we even should not make it an end user documentation and it is not the matter of our opinions but I looked it up again and this is rather explicitly specified int the ASF release policy: 

https://www.apache.org/legal/release-policy.html#publication

> Projects SHALL publish official releases and SHALL NOT publish unreleased materials outside the development community.

> During the process of developing software and preparing a release, various packages are made available to the development community for testing purposes. Projects MUST direct outsiders towards official releases rather than raw source repositories, nightly builds, snapshots, release candidates, or any other similar packages. Projects SHOULD make available developer resources to support individuals actively participating in development or following the dev list and thus aware of the conditions placed on unreleased materials.

So really patching 'main' should not be advertised to anyone outside of the development community (so anyone who uses GitHub and reads devlist. 

And this has strong foundation from the legal point of view I believe. The release process where release is published as explicit 'act' of PMC has certIn legal implications. And anyone using those might be aware of the conditions. This is really about potential litigations against the ASF or PMC members.  If someone loses a fortune using released sources and PMC voted on this and all was according to ASF policies, then ASF provides idem ification to PMC members and take over the litigation burden (or most likely just assist with dropping the case). But this is under the condition that the user uses released code. If we are suggesting 'main' in the user docs, we effectively bypass this. If we discuss it in GitHub issues/PRs/devlist this is a different story because the context of this is clearly 'devlopment' community. If we put it in the user documentation, there is no such distinction.

Reply:
@potiuk i think `publish` is a key word there.  is providing instructions for applying a patch of your choosing the same thing as publishing unreleased materials?  i'm not a lawyer; but i really don't think it's the same thing.  that seems rather like saying that providing instructions for how to `pip install` a dependency is "publishing" a dependency.  additionally, specific reference to main is problematic, we could rephrase to simply provide instructions for applying a PR or a commit, without specifying what branch that commit may be part of.  it's not just patching unreleased code that is useful here; you might want to do your own backporting.

Reply:
anyway, closing it

Reply:
> just patching unreleased code that is useful here; you might want to do your own backporting

Yeah. But "airflow.apache.org" should only be about the code that is released via "https://downloads.apache.org/" - this is the only way ASF releases "end-user" software. We also release "Docker Images" and "PIP" packages as "convenience/compiled packages" but - also those https://www.apache.org/legal/release-policy.html#compiled-packages:

> As a convenience to users that might not have the appropriate tools to build a compiled version of the source, binary/bytecode packages MAY be distributed alongside official Apache releases. In all such cases, the binary/bytecode package MUST have the same version number as the source release and MUST only add binary/bytecode files that are the result of compiling that version of the source code release and its dependencies.

So technically speaking - I think we cannot - in the "end-user" documentation provide the user instruction how to patch using PR/main/commits that have not been released yet. This is a bit blurry of course. 

But I think - to be honest - having those instructions are STILL valuable for the "power users" who are part of the "development community" - but not "end-users".  This is mostly to reflect ASF philosophy and legal requirements, not technical reasons. 

I am actually curious if my way of thinking is correct or not. I will ask a LEGAL JIRA issue about it (this is the way how we can get the ASF statement about this actually and we will stop having opinions - both of us - we can get either some legal ASF team opinion on that, or even authoritative answer.

Reply:
I created a legal issue: https://issues.apache.org/jira/browse/LEGAL-621 let's see what the legal team at the ASF has to say.

Reply:
sounds good, thanks 👍

Reply:
Right. I think I got some interesting feedback in the https://issues.apache.org/jira/browse/LEGAL-621 and I believe we should have a section on our page that will be clearly marked as "For developers" / "Deveoloper tools" where we make appropriate disclaimers and where we will not only provide such instructions but also add some more information that might direct people who want to start messing with their Airlfow code to our Github repository.

Seems like other projects do that (Spark for example).

This might go hand-in-hand with the discussion with slack we have about restructuring our documentation: https://apache-airflow.slack.com/archives/CJ1LVREHX/p1661234606247079  and seems to be in-line with the ASF approach, and it would remove the need for test automation. Just having a clear, separate section that you will know it's not for "regular/non-developer" users and have enough disclaimers so that people expect that it MAY work rather than HOULD work. 

Comments from the LEGAL JIRA of mine:

> [Sean R. Owen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srowen)  - I like your example about Spark.

> We are also about to redo the structure of our documentation (we have an open discussion about it) and I will propose to add a new section "Developer tools", where it will be clearly explained that there are some ways developers engaged in our community developemnt can do stuff and I think adding it in the "airflow.apache.org" documentation is also fine - as long as we have a clear "disclaimer" for this section. Similarly we have https://airflow.apache.org/ecosystem/ ecosystem page where we (According to ASF policies I believe) have links to stuff which comes from ecosystem but is not "Apache" software and we have clear disclaimers about that).

> If this is an acceptable approach for others from logal looking at it, then I think that solves our discussion and we can close the ticket, I will bring it to the PMC and we will likely go that route.

> Such approach is nice because it might also provide a nice communication tool targeted to our users who would like to become contributors. we can add there some encouragement for those who are both users and Python developers to take a look at our Github Repo, so we can make it another way of "Growing contributors base".

Reply:
Also in https://issues.apache.org/jira/browse/LEGAL-621?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17611927#comment-17611927 there is comment from Craig about making clear distinction between "end users" and "developers" (that's where my initial doubts also came from) which I think we have to be very clear in the "developer tools" section about.